### ISSUE TITLE
[REVIEW]: AtomNeb: IDL Library for Atomic Data of Ionized Nebulae

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#898]
**Submitting author:** @danehkar (<a href="http://orcid.org/0000-0003-4552-5997">Ashkbiz Danehkar</a>)
**Repository:** <a href="https://github.com/atomneb/AtomNeb-idl" target ="_blank">https://github.com/atomneb/AtomNeb-idl</a>
**Version:** v0.0.2
**Editor:** @arfon
**Reviewer:** @mgalloy, @mdpiper
**Archive:** <a href="https://doi.org/10.5281/zenodo.2584420" target="_blank">10.5281/zenodo.2584420</a>

## Status

[![status](http://joss.theoj.org/papers/7205fc05ec14970b77c4551c2fe6b0bf/status.svg)](http://joss.theoj.org/papers/7205fc05ec14970b77c4551c2fe6b0bf)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/7205fc05ec14970b77c4551c2fe6b0bf"><img src="http://joss.theoj.org/papers/7205fc05ec14970b77c4551c2fe6b0bf/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/7205fc05ec14970b77c4551c2fe6b0bf/status.svg)](http://joss.theoj.org/papers/7205fc05ec14970b77c4551c2fe6b0bf)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/atomneb/AtomNeb-idl" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@mgalloy & @mdpiper, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.theoj.org/about#reviewer_guidelines. Any questions/concerns please let @arfon know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/atomneb/AtomNeb-idl">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** v0.0.2
- [x] **Authorship:** Has the submitting author (@danehkar) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/atomneb/AtomNeb-idl">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** v0.0.2
- [x] **Authorship:** Has the submitting author (@danehkar) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @mgalloy, it looks like you're currently assigned as the reviewer for this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[--> Check article proof :page_facing_up: <--](https://github.com/openjournals/joss-papers/blob/joss.00898/joss.00898/10.21105.joss.00898.pdf)

**Comment by mdpiper** [comment]
Info for contributors (last item in the **Documentation** section) is requested in https://github.com/atomneb/AtomNeb-idl/issues/1.

**Comment by mdpiper** [comment]
The package isn't stamped with a release, so I created https://github.com/atomneb/AtomNeb-idl/issues/3 to request one. This addresses the third item in the **General checks** checklist.

**Comment by mdpiper** [comment]
The package doesn't include API documentation (the fourth item in the **Documentation** checklist). I've requested this in https://github.com/atomneb/AtomNeb-idl/issues/4.

**Comment by mdpiper** [comment]
I offered a PR (https://github.com/atomneb/AtomNeb-idl/pull/5) to automatically locate the package data from the examples.

**Comment by mdpiper** [comment]
Created an issue (https://github.com/atomneb/AtomNeb-idl/issues/6) to address package installation (the first item in the **Functionality** checklist).

**Comment by mdpiper** [comment]
My review of **AtomNeb** is complete. The library and its associated paper are appropriate for publication in JOSS, pending  @danehkar's responses to the minor issues I've opened in the repository.

(@arfon I'll complete the open items in the review checklist once these issues have been addressed.)

**Comment by arfon** [comment]
Thanks @mdpiper!

@mgalloy - how are you getting along here?

**Comment by danehkar** [comment]
@mdpiper thank you for your reviews. I will solve those minor issues. did you also finish your review of **proEQUIB**?

**Comment by mgalloy** [comment]
Catching up after some travel. Added some comments on issues opened by @mdpiper.

I think it would also be useful to add a short sentence after the opening description of AtomNeb database in your [README](https://github.com/atomneb/AtomNeb-idl/blob/master/README.md) that indicates precisely how `AtomNeb-idl` relates to it. Does it contain the data, access them via the web, access data you have generated, etc.?

I should be able to start looking at proEQUIB shortly.

**Comment by arfon** [comment]
@danehkar - I think this review is waiting on you at this point?

**Comment by danehkar** [comment]
> @mdpiper Info for contributors (the last item in the **Documentation** checklist) isn't present, so I made a request in [atomneb/AtomNeb-idl#1](https://github.com/atomneb/AtomNeb-idl/issues/1).

I included [CONTRIBUTING.md](https://github.com/atomneb/AtomNeb-idl/blob/master/CONTRIBUTING.md) in the [AtomNeb-idl](https://github.com/atomneb/AtomNeb-idl) repository that contains some guidelines for contributors. Now I closed the issue [atomneb/AtomNeb-idl#1](https://github.com/atomneb/AtomNeb-idl/issues/1).

**Comment by danehkar** [comment]
> @mdpiper I offered a PR ([atomneb/AtomNeb-idl#5](https://github.com/atomneb/AtomNeb-idl/pull/5)) to automatically locate the package data from the examples.

Thank you! I merged your PR into the repository. Now the package data are automatically located in the examples: [atomneb/AtomNeb-idl/examples](https://github.com/atomneb/AtomNeb-idl/tree/master/examples).

**Comment by danehkar** [comment]
> @mdpiper Created an issue ([atomneb/AtomNeb-idl#6](https://github.com/atomneb/AtomNeb-idl/issues/6)) to address package installation (the first item in the **Functionality** checklist).

Thank you for your installation note suggestion. I added an installation note in [README.md](https://github.com/atomneb/AtomNeb-idl/blob/master/README.md):

> To install the **AtomNeb** database and its API library, simply add the path of this package directory to your IDL path. This package requires IDL version 7.1 or later. For more information about the path management in IDL, read [the tips for customizing IDL program path](https://www.harrisgeospatial.com/Support/Self-Help-Tools/Help-Articles/Help-Articles-Detail/ArtMID/10220/ArticleID/16156/Quick-tips-for-customizing-your-IDL-program-search-path) provided by Harris Geospatial Solutions or [the IDL library installation note](http://www.idlcoyote.com/code_tips/installcoyote.php) by David Fanning in the Coyote IDL Library.  

Now I closed the issue [atomneb/AtomNeb-idl#6](https://github.com/atomneb/AtomNeb-idl/issues/6).

**Comment by mdpiper** [comment]
@danehkar With your recent updates, I've marked as complete a few more items on the checklist.

**Comment by danehkar** [comment]
@labarba I am currently working on those issues. Hopefully, I will resolve them in a month. Thank you for your patience.

**Comment by danehkar** [comment]
> The package doesn't include API documentation (the fourth item in the **Documentation** checklist). I've requested this in [atomneb/AtomNeb-idl#4](https://github.com/atomneb/AtomNeb-idl/issues/4).

@mdpiper The documentation is now hosted at https://atomneb.github.io/AtomNeb-idl/doc/

**Comment by danehkar** [comment]
> This package doesn't include unit tests (second-to-last item in the **Documentation** checklist). I've made a request for testing with [atomneb/AtomNeb-idl#2](https://github.com/atomneb/AtomNeb-idl/issues/2).

 @mdpiper I added  [mgunit](https://github.com/mgalloy/mgunit) in the folder  [test_mgunit](https://github.com/atomneb/AtomNeb-idl/tree/master/test_mgunit) for automated tests. 

To test, you can download AtomNeb-idl with its dependencies as follows 

    git clone --recursive https://github.com/atomneb/AtomNeb-idl.git

You need to unpack rc_o_iii_SSB17.fits.tar.gz in ``atomic-data-rc``:

    tar -xvf rc_o_iii_SSB17.fits.tar.gz

To run the test, you need to run the following command:

    idl test_all.pro

There should not be any errors in ``test-results.html``.

**Comment by mgalloy** [comment]
I ran the tests and confirm they all passed. One note: I had to go get `mgutlibtestcase__define.pro` — it is not part of `mgunit`. I would add it to the `test_mgunit` directory, possibly renaming to `atomnebuttestcase__define.pro`.

**Comment by danehkar** [comment]
> I ran the tests and confirm they all passed. One note: I had to go get `mgutlibtestcase__define.pro` — it is not part of `mgunit`. I would add it to the `test_mgunit` directory, possibly renaming to `atomnebuttestcase__define.pro`.

Thank you for testing. I will copy `mgutlibtestcase__define.pro`to the `test_mgunit` directory, and rename it  to `atomnebuttestcase__define.pro`.

**Comment by danehkar** [comment]
@arfon I would like to rename the _title_ of this submission ([898](https://github.com/openjournals/joss-reviews/issues/898)) from **AtomNeb: Atomic Data for Ionized Nebulae** to **AtomNeb: IDL Library for Atomic Data of Ionized Nebulae** since I am working on a similar python package, and I want to distinguish between them. I will upload the revised paper of this IDL library soon.

**Comment by arfon** [comment]
> @arfon I would like to rename the _title_ of this submission ([898](https://github.com/openjournals/joss-reviews/issues/898)) from **AtomNeb: Atomic Data for Ionized Nebulae** to **AtomNeb: IDL Library for Atomic Data of Ionized Nebulae** since I am working on a similar python package, and I want to distinguish between them. I will upload the revised paper of this IDL library soon.

OK, no problem. I've updated the issue title here.

**Comment by danehkar** [comment]
> I ran the tests and confirm they all passed. One note: I had to go get `mgutlibtestcase__define.pro` — it is not part of `mgunit`. I would add it to the `test_mgunit` directory, possibly renaming to `atomnebuttestcase__define.pro`.

@mgalloy Thank you for the correction. I added it to `test_mgunit` directory.

**Comment by mgalloy** [comment]
OK, assuming you tag a version after you have addressed any problems from the other reviewers, my review is complete.

**Comment by danehkar** [comment]
@arfon I updated the paper, and regenerated the pdf.  You can check it here: [799](https://github.com/openjournals/joss-reviews/issues/799).

**Comment by arfon** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.00898/joss.00898/10.21105.joss.00898.pdf)

**Comment by danehkar** [comment]
@mgalloy Thank you very much for your test and review of this package. I look forward to @mdpiper's comments. 

@arfon Thank you for regenerating the article proof. It looks fine.

**Comment by mdpiper** [comment]
@danehkar With your recent updates, I've marked off the last three items on my checklist, completing my review. One minor issue, which I defer to @arfon, is the version on the software (v0.0.1) doesn't quite follow semantic versioning. If this isn't an issue for JOSS, it isn't for me either.

**Comment by arfon** [comment]
> @danehkar With your recent updates, I've marked off the last three items on my checklist, completing my review. One minor issue, which I defer to @arfon, is the version on the software (v0.0.1) doesn't quite follow semantic versioning. If this isn't an issue for JOSS, it isn't for me either.

We can update the version at the end of the review if needed. Just let me know what version you're reviewing. 

Also, we don't insist on using semantic versioning.

**Comment by danehkar** [comment]
Considering Semantic Versioning (**MAJOR.MINOR.PATCH**), the code did not have any MAJOR change such as incompatible API changes, and any MINOR changes (e.g. adding a new functionality) since it was submitted. It has some changes in comments for documentation, so I changed the PATCH version from v0.0.1 to v0.0.2. I added an MGUnit tester, while the program functions were not changed, so there are no major or minor changes.

@arfon Please update Version from v0.0.1 to v0.0.2.

**Comment by arfon** [comment]
@whedon set v0.0.2 as version

**Comment by whedon** [comment]
OK. v0.0.2 is the version.

**Comment by arfon** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.00898/joss.00898/10.21105.joss.00898.pdf)

**Comment by arfon** [comment]
@danehkar - At this point could you make an archive of the reviewed software in Zenodo/figshare/other service and update this thread with the DOI of the archive? I can then move forward with accepting the submission.

**Comment by danehkar** [comment]
@arfon I did the proof correction in the paper. You can check it here: [799](https://github.com/openjournals/joss-reviews/issues/799). Please regenerated the final pdf for the paper. I also made an archive of the reviewed software in Zenodo here: https://doi.org/10.5281/zenodo.2584420

**Comment by arfon** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.00898/joss.00898/10.21105.joss.00898.pdf)

**Comment by arfon** [comment]
@whedon set 10.5281/zenodo.2584420 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.2584420" target="_blank">10.5281/zenodo.2584420</a> is the archive.

**Comment by arfon** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/539

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/539, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.1086/498148 is OK
- 10.1051/0004-6361/200911712 is OK
- 10.1088/0004-637X/744/2/99 is OK
- 10.1051/0004-6361/201323152 is OK
- 10.1051/aas:2000139 is OK
- 10.1086/191479 is OK
- 10.1093/mnras/272.2.369 is OK
- 10.1051/aas:1998319 is OK
- 10.1093/mnras/272.1.41 is OK
- 10.1111/j.1745-3933.2012.01300.x is OK
- 10.1093/mnrasl/slt049 is OK
- 10.1051/0004-6361/201116511 is OK
- 10.1051/0004-6361/201116511e is OK
- 10.1093/mnras/stx1189 is OK
- 10.1093/mnras/stt1116 is OK
- 10.1093/mnras/stu203 is OK
- 10.3847/0004-6256/151/2/38 is OK
- 10.1017/pasa.2018.1 is OK
- 10.21105/joss.00899 is OK
- 10.1086/316190 is OK
- 10.1086/319184 is OK
- 10.1046/j.1365-8711.2003.06371.x is OK
- 10.1111/j.1365-2966.2005.09381.x is OK
- 10.1086/524378 is OK
- 10.1051/0004-6361:20010923 is OK
- 10.1051/0004-6361/201015362 is OK

MISSING DOIs

- https://doi.org/10.1051/0004-6361:20020263 may be missing for title: Recombination coefficients for O II lines at nebular temperatures and densities
- https://doi.org/10.1016/0032-0633(92)90127-a may be missing for title: Total and effective radiative recombination coefficients
- https://doi.org/10.1117/12.959819 may be missing for title: FITS - a Flexible Image Transport System

INVALID DOIs

- None
```

**Comment by arfon** [comment]
@danehkar - could you check those potentially missing DOIs and see if they are the correct ones for some of your references?

**Comment by danehkar** [comment]
@arfon Those references with missing DOI are from **Astronomy & Astrophysics** and **Astronomy and Astrophysics Supplement**:
[1994A&A...282..999S](http://adsabs.harvard.edu/abs/1994A%26A...282..999S)
[1991A&A...251..680P](http://adsabs.harvard.edu/abs/1991A%26A...251..680P)
[1981A&AS...44..363W](http://adsabs.harvard.edu/abs/1981A%26AS...44..363W)
These two journals (**Astronomy & Astrophysics** and **Astronomy and Astrophysics Supplement**) did not assign any Digital Object Identifier (DOI) to their publications prior to 2001, and they still have not assigned any DOI to their papers published before 2001.

Those DOI identified by the robot **whedon** are NOT related to these 3 references:
- https://doi.org/10.1051/0004-6361:20020263 may be missing for title: Recombination coefficients for O II lines at nebular temperatures and densities
- https://doi.org/10.1016/0032-0633(92)90127-a may be missing for title: Total and effective radiative recombination coefficients
- https://doi.org/10.1117/12.959819 may be missing for title: FITS - a Flexible Image Transport System

I confirm that the references in `paper.bib` are correct, as they are BibTex entries from [adsabs.harvard.edu](http://adsabs.harvard.edu/).

**Comment by arfon** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/540
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.00898](https://doi.org/10.21105/joss.00898)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by arfon** [comment]
@mgalloy, @mdpiper - many thanks for your reviews here ✨ 

@danehkar - your paper is now accepted into JOSS :zap::rocket::boom:

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](http://joss.theoj.org/papers/10.21105/joss.00898/status.svg)](https://doi.org/10.21105/joss.00898)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.00898">
  <img src="http://joss.theoj.org/papers/10.21105/joss.00898/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: http://joss.theoj.org/papers/10.21105/joss.00898/status.svg
   :target: https://doi.org/10.21105/joss.00898
```
This is how it will look in your documentation:

[![DOI](http://joss.theoj.org/papers/10.21105/joss.00898/status.svg)](https://doi.org/10.21105/joss.00898)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: http://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://www.flipcause.com/secure/cause_pdetails/Mjk3Nzk=
